controllers/bucket_controller.go

```diff
@@ -82,403 +82,471 @@ func (r *BucketReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts Buc
-	}
-	if r.ExternalEventRecorder != nil {
-		objRef, err := reference.GetReference(r.Scheme, &bucket)
-		if err != nil {
-			log.Error(err, "unable to send event")
-			return
+// buildClient constructs a minio.Client with the data from the given object and secret.
+// It returns an error if the given Secret does not have the required fields, or if there is no credential handler
+// configured.
+func (r *BucketReconciler) buildClient(obj *sourcev1.Bucket, secret *corev1.Secret) (*minio.Client, error) {
+	opts := minio.Options{
+		Region: obj.Spec.Region,
+		Secure: !obj.Spec.Insecure,
+	}
+	if secret != nil {
+		var accessKey, secretKey string
+		if k, ok := secret.Data["accesskey"]; ok {
+			accessKey = string(k)
 		}
-
-		if err := r.ExternalEventRecorder.Eventf(*objRef, nil, severity, severity, msg); err != nil {
-			log.Error(err, "unable to send event")
-			return
+		if k, ok := secret.Data["secretkey"]; ok {
+			secretKey = string(k)
 		}
+		if accessKey == "" || secretKey == "" {
+			return nil, fmt.Errorf("invalid '%s' secret data: required fields 'accesskey' and 'secretkey'", secret.Name)
+		}
+		opts.Creds = credentials.NewStaticV4(accessKey, secretKey, "")
+	} else if obj.Spec.Provider == sourcev1.AmazonBucketProvider {
+		opts.Creds = credentials.NewIAM("")
 	}
-}
 
-func (r *BucketReconciler) recordReadiness(ctx context.Context, bucket sourcev1.Bucket) {
-	log := logr.FromContext(ctx)
-	if r.MetricsRecorder == nil {
-		return
-	}
-	objRef, err := reference.GetReference(r.Scheme, &bucket)
-	if err != nil {
-		log.Error(err, "unable to record readiness metric")
-		return
-	}
-	if rc := apimeta.FindStatusCondition(bucket.Status.Conditions, meta.ReadyCondition); rc != nil {
-		r.MetricsRecorder.RecordCondition(*objRef, *rc, !bucket.DeletionTimestamp.IsZero())
-	} else {
-		r.MetricsRecorder.RecordCondition(*objRef, metav1.Condition{
-			Type:   meta.ReadyCondition,
-			Status: metav1.ConditionUnknown,
-		}, !bucket.DeletionTimestamp.IsZero())
-	}
+	return minio.New(obj.Spec.Endpoint, &opts)
 }
 
-func (r *BucketReconciler) recordSuspension(ctx context.Context, bucket sourcev1.Bucket) {
-	if r.MetricsRecorder == nil {
-		return
+// revision calculates the SHA256 checksum of the given string map.
+// The keys are sorted to ensure a stable order, and the SHA256 sum is then calculated for the string representations of
+// the key/value pairs, each pair written on a newline. The sum result is returned as a string.
+func (r *BucketReconciler) revision(list map[string]string) string {
+	keyIndex := make([]string, 0, len(list))
+	for k := range list {
+		keyIndex = append(keyIndex, k)
 	}
-	log := logr.FromContext(ctx)
-
-	objRef, err := reference.GetReference(r.Scheme, &bucket)
-	if err != nil {
-		log.Error(err, "unable to record suspended metric")
-		return
+	sort.Strings(keyIndex)
+	sum := sha256.New()
+	for _, k := range keyIndex {
+		sum.Write([]byte(fmt.Sprintf("%s  %s\n", k, list[k])))
```

Comment:
`Write()` returns an int and an error. Is the error intentionally ignored here?

Reply:
There is not a scenario in which `Write` would return an error, but you are right about taking it into account in case we change implementation.